### PR TITLE
Add 2 methods: GetListenPort() and SessionStats()

### PR DESCRIPTION
**Pull request by liut**
new two methods of DelugeClient:
```go

GetListenPort() (uint16, error)
SessionStats() (*SessionStats, error)

```

They depend on [go-rencode PR 7](https://github.com/gdm85/go-rencode/pull/7).

And:
* Use mutual exclusion lock to fix mismatch c.serial in goroutines.

### CODE REVIEW

File: methods.go [hunk 1]
```diff
@@ -42,6 +42,20 @@ func (c *Client) GetFreeSpace(path string) (int64, error) {
 	return freeSpace, nil
 }
 
+// GetListenPort return the listen port of daemon session
+func (c *Client) GetListenPort() (port uint16, err error) {
```
Comment (gdm85) [inline methods.go:46]: Please do not use named return values, they are not used anywhere else.

File: methods.go [hunk 2]
```diff
@@ -42,6 +42,20 @@ func (c *Client) GetFreeSpace(path string) (int64, error) {
 	return freeSpace, nil
 }
 
+// GetListenPort return the listen port of daemon session
```
Comment (gdm85) [inline methods.go:45]:
```suggestion
// GetListenPort return the listen port of the Deluge daemon.
```

File: methods.go [hunk 3]
```diff
@@ -42,6 +42,20 @@ func (c *Client) GetFreeSpace(path string) (int64, error) {
...
+func (c *Client) GetListenPort() (port uint16, err error) {
+	var resp *DelugeResponse
+	resp, err = c.rpc("core.get_listen_port", rencode.List{}, rencode.Dictionary{})
+	if err != nil {
+		return
+	}
+	if err != nil {
```
Comment (gdm85) [inline methods.go:52]: This is incorrect, see any other method.

File: session_stats.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package delugeclient
+
+import (
+	"github.com/gdm85/go-rencode"
+)
+
+// SessionStats ...
```
Comment (gdm85) [inline session_stats.go:7]: ...?

File: session_stats.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+type SessionStats struct {
+	PayloadDownloadRate int64
+	PayloadUploadRate   int64
+	DownloadRate        int64
+	UploadRate          int64
+	TotalDownload       int64 // net.recv_bytes
+	TotalUpload         int64 // net.sent_bytes
+	// FreeSpace           int64
```
Comment (gdm85) [inline session_stats.go:15]: Why are these commented?

File: session_stats.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	"total_upload",
+	// "free_space",
+	"dht_nodes",
+	"num_peers",
+	// "num_torrents",
+	// "num_paused_torrents",
+	"has_incoming_connections",
+)
+
+// SessionStats ...
```
Comment (gdm85) [inline session_stats.go:38]: ?

File: session_stats.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		args.Add(rencode.NewList(keys))
+	}
+
+	rd, err := c.rpcWithDictionaryResult("core.get_session_status", args, rencode.Dictionary{})
+	if err != nil {
+		return nil, err
+	}
+
+	if c.settings.Logger != nil {
+		c.settings.Logger.Printf("got stats: %+s", rd)
```
Comment (gdm85) [inline session_stats.go:53]:
```suggestion
		c.settings.Logger.Printf("session stats: %+s", rd)
```

File: torrent_status.go [hunk 8]
```diff
@@ -75,6 +75,10 @@ const (
 	StateMoving      TorrentState = "Moving"
 )
 
+func (ts TorrentStatus) GetState() TorrentState {
```
Comment (gdm85) [inline torrent_status.go:78]: GoDoc missing here

File: session_stats.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	// "num_torrents",
+	// "num_paused_torrents",
+	"has_incoming_connections",
+)
+
+// SessionStats ...
+func (c *Client) SessionStats(keys ...string) (*SessionStats, error) {
+	var args rencode.List
+	if 0 == len(keys) {
+		args.Add(sessStatsKeys)
```
Comment (gdm85) [inline session_stats.go:42]: What happens when no keys are passed instead?

**Review by gdm85 — CHANGES_REQUESTED** [review]

**Comment by gdm85** [comment]
Thanks for your PR! Left a review.

File: session_stats.go (hunk 5 above)
Comment (liut) [reply session_stats.go:15]: These fields read failed, I'll fix it soon

File: methods.go (hunk 1 above)
Comment (liut) [reply methods.go:46]: Oh, I see.

File: session_stats.go (hunk 9 above)
Comment (liut) [reply session_stats.go:42]: It inspired from deluge.
The process in deluge.core.core like this:
```python
        if not keys:
            return self.session_status

        status = {}
        for key in keys:
                status[key] = self.session_status[key]
        return status
```

But actually these are not many and the whole return is Ok, fix it later.

**Comment by liut** [comment]
@gdm85  Thanks for you reply, the PR will change soon, Happy weekend!

File: delugeclient.go [hunk 10]
```diff
@@ -292,6 +296,9 @@ func (c *Client) Close() error {
 const Deluge2ProtocolVersion = 1
 
 func (c *Client) rpc(methodName string, args rencode.List, kwargs rencode.Dictionary) (*DelugeResponse, error) {
+	c.mu.Lock()
```
Comment (gdm85) [inline delugeclient.go:299]: Why is this added here?

Comment (gdm85) [reply delugeclient.go:299]: I read from your description:
> Use mutual exclusion lock to fix mismatch c.serial in goroutines.

But this seems unrelated to adding the two methods and a pretty big change. What is the reason you added this? What did you discover about the Deluge Daemon RPC?

**Comment by gdm85** [comment]
There are no tests, please see how other methods have tests for them (unit test and using fixture bytes)

File: methods.go [hunk 11]
```diff
@@ -42,6 +42,23 @@ func (c *Client) GetFreeSpace(path string) (int64, error) {
 	return freeSpace, nil
 }
 
+// GetListenPort return the listen port of the deluge daemon
+func (c *Client) GetListenPort() (uint16, error) {
```
Comment (gdm85) [inline methods.go:46]: Also, this function is missing from the interface.

**Comment by gdm85** [comment]
I have implemented `GetListenPort()` in a0b94c9; please see if it works well for you and you can use a similar approach to adjust the code for `SessionStats()` here in this PR.

File: delugeclient.go (hunk 10 above)
Comment (liut) [reply delugeclient.go:299]: Deluge Daemon is OK. But I found a lot of error from client log likes:
```
request/response serial id mismatch: got 886067 but 886070 expected
```
I have 2-3 goroutines to read response from one Client instance at the same time, but the time spent of responses are not consistent, I think here is why mismatch serial number.
No better solution for now.
btw more tests will up later.

File: methods.go (hunk 11 above)
Comment (liut) [reply methods.go:46]: Yes, it have.

Comment (liut) [reply methods.go:46]: Yes, it has. and I will check out it for merge from main branch.

File: delugeclient.go (hunk 10 above)
Comment (liut) [reply delugeclient.go]: Yes, it's another question.

**Comment by ludviglundgren** [comment]
Any updates on `SessionStats()`? This would be very useful for multiple projects of mine.

File: delugeclient.go [hunk 12]
```diff
@@ -293,6 +294,7 @@ func (c *Client) Close() error {
 const Deluge2ProtocolVersion = 1
 
 func (c *Client) rpc(methodName string, args rencode.List, kwargs rencode.Dictionary) (*DelugeResponse, error) {
+
```
Comment (gdm85) [inline delugeclient.go:297]:
```suggestion
```

File: session_stats.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
+package delugeclient
+
+import (
+	"github.com/gdm85/go-rencode"
+)
+
+// SessionStats Basic session stats
```
Comment (gdm85) [inline session_stats.go:7]:
```suggestion
// SessionStats contains basic session statistics.
```

File: session_stats.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+	DownloadRate           int64
+	UploadRate             int64
+	TotalDownload          int64
+	TotalUpload            int64
+	DhtNodes               int16
+	NumPeers               int16
+	HasIncomingConnections bool
+}
+
+// sessStatsKeys default keys of session stats, see "deluge/core/core.py"
```
Comment (gdm85) [inline session_stats.go:20]:
```suggestion
// sessStatsKeys is a slice with all the default keys for session stats.
```

File: session_stats.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	"download_rate",
+	"upload_rate",
+	"total_download",
+	"total_upload",
+	"dht_nodes",
+	"num_peers",
+	"has_incoming_connections",
+)
+
+// SessionStats Gets session status values as SessionStats object
```
Comment (gdm85) [inline session_stats.go:33]:
```suggestion
// SessionStats retrieves session statistics from the session status.
```

File: torrent_status.go [hunk 16]
```diff
@@ -75,6 +75,11 @@ const (
 	StateMoving      TorrentState = "Moving"
 )
 
+// GetState return State string value as TorrentState type
```
Comment (gdm85) [inline torrent_status.go:78]:
```suggestion
// GetState returns State string value as TorrentState type.
```

**Review by gdm85 — APPROVED** [review]

**Comment by gdm85** [comment]
@ludviglundgren I never received a new review request, although now I see the PR is almost complete. The CI is failing right now.

**Comment by ludviglundgren** [comment]
Ah cool! 

It errors on:

```
 === CONT  TestSessionStats
    methods_test.go:189: field "PayloadDownloadRate": value 0: cannot convert from float32 into *int64
```

Is this something you can possibly take a look at if you have time? I'm not really familiar with the codebase, but if you have your plate full I might have to do it. 🙂

**Comment by liut** [comment]
This PR depend on [go-rencode#7](https://github.com/gdm85/go-rencode/pull/7)

**Comment by gdm85** [comment]
@liut why not use here the float as it is returned from the RPC response?

**Comment by liut** [comment]
`PayloadDownloadRate` is not a float32.

https://github.com/arvidn/libtorrent/blob/RC_1_2/include/libtorrent/torrent_status.hpp#L327

**Comment by gdm85** [comment]
They are `float` here: http://svn.netlabs.org/repos/ports/libtorrent/vendor/0.14.12/docs/manual.html

**Comment by gdm85** [comment]
Superseded by #39
